Use explicit destinations in codegen to avoid uselessly jumping around.

sjrd opened the pull request:
Backport of https://github.com/lampepfl/dotty/pull/14890

sjrd committed:
Add bytecode tests with the status quo of codegen control flow.

sjrd committed:
Remove a fallback in InlinerTest.oldInlineHigherOrderTest().

sjrd committed:
Use explicit destinations in codegen to avoid uselessly jumping around.

Previously, the codegen's main method `genLoad` always generated
code that loaded the value on the stack before continuing. There
were a number of situations where `genLoad` would be directly
followed by unconditional jumps to instructions performing more
jumps, returns and throws. This generated more spurious jumps than
necessary, along with artifact dead code.

We solve these limitations by introducing `LoadDestination`s that
specify the destination of a loaded value:

* FallThrough: as previously, load the value on the stack and continue.
* Jump(label): load the value on the stack and jump to the given label.
* Return: return the value from the enclosing method.
* Throw: throw the value.

We generalize `genLoad` as `genLoadTo`, taking a specific
destination for the loaded value. `genLoadTo` can "push down" its
destination into all control flow structures (except `Try`s,
because of their cleanups). With that, when we get to the end of
what amounts to "basic blocks", we know exactly the ultimate
destination of the loaded value. We can therefore directly jump,
return or throw to the final destination.

This produces less bytecode, notably because fewer labels are
necessary. For example, the method:

  def abs(x: Int): Int = if x < 0 then -x else x

previously generated bytecode like

  ILOAD 1
  ICONST_0
  IF_ICMPGE Label(1)
  ILOAD 1
  INEG
  GOTO Label(2)
  Label(1):
  ILOAD 1
  Label(2):
  IRETURN

Now, instead of jumping to Label(2), we directly perform an
IRETURN:

  ILOAD 1
  ICONST_0
  IF_ICMPGE Label(1)
  ILOAD 1
  INEG
  IRETURN
  Label(1):
  ILOAD 1
  IRETURN

While the changes are not very impressive on that simple example,
they become more important in more complex cases, notably with
switch matches. Examples can be found in the changed bytecode
tests.

An added benefit is that `genLoadTo` knows when loading a value
results in an unconditional control flow change (jump, return or
throw). It can then avoid inserting any useless adaptation. This
removes all the dead bytecode that the codegen used to generate
as artifacts of its own compilation scheme. (It will still
generate dead bytecode if the original source code/inlined code
contains dead code.)

This is a backport of dotty's commit
https://github.com/lampepfl/dotty/commit/4a2889f93a46372c3551beef8b17d4ba8f289ddf

The improvements in scalac are not as good as in dotc because
jumps to `LabelDef`s cannot be tracked in the same way as the
returns from `Labeled` blocks. For the latter, by the time we see
a return from a labeled block, we always already know the ultimate
destination of the labeled block itself, and we can therefore jump
to it. In scalac, we can find jumps to `LabelDef`s before knowing
their ultimate destination. In this commit, we therefore do not
push down the jumps to `LabelDef`s as `LoadDestination`s. We leave
that to a separate commit, to keep this commit as close as possible
to the dotty commit it backports.

lrytz commented:
So nice, thank you!

sjrd commented on test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala:
```diff
@@ -1443,17 +1443,11 @@ class InlinerTest extends BytecodeTesting {
-        -1 /*A*/, "$anonfun$t$3", IRETURN))
-      catch { case _: AssertionError => throw e }
-    }
+      "$anonfun$t$2", ISTORE, GOTO /*B*/,
+      -1 /*A*/, "$anonfun$t$3", ISTORE,
+      -1 /*B*/, ILOAD, IRETURN))
```
@lrytz Do you have any idea why the optimizer generates *worse* code after these changes than before? With the changes, with or without optimizer, the bytecode summary generated for `ifelse` is
```
ALOAD, "apply$mcZ$sp", IFEQ /*8*/, ALOAD, "apply", ARETURN, -1 /*8*/, ALOAD, "apply", ARETURN
```
So why does the generated code for `t` regresses to having an `ISTORE`/`GOTO`/`ILOAD`?

Or perhaps a better question: how did the optimizer have the direct `IRETURN`s *before* the changes?

Do you have any pointer in the optimizer to help me figure this out?

lrytz replied:
It seems the difference in the generated code is that the `store` is pushed into the two branches. before there were jumps to the single store instruction.

`sc12 Test.scala -opt:l:inline '-opt-inline-from:**' -Yopt-trace C.t`


```diff
@@ -65,27 +65,27 @@
     POP
     INVOKESTATIC C.$anonfun$t$2 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
+    ASTORE 5
     GOTO L9
    L7
     ALOAD 4
     POP
     INVOKESTATIC C.$anonfun$t$3 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
-   L9
     ASTORE 5
-    GOTO L10
-   L11
+    GOTO L9
    L10
+   L9
     ALOAD 5
-    LINENUMBER 4 L10
+    LINENUMBER 4 L9
     INVOKESTATIC scala/runtime/BoxesRunTime.unboxToInt (Ljava/lang/Object;)I
     IRETURN
-   L12
+   L11
```

So the store-load-pair removal no longer triggers

https://github.com/scala/scala/blob/7dbc139b4a113c18e99e8720b985a249fc9b181f/src/compiler/scala/tools/nsc/backend/jvm/opt/CopyProp.scala#L476-L500

sjrd replied:
Thanks. That sounds plausible indeed! I'll see if I can do something to recover from that.

sjrd replied:
I pushed a commit in the inliner that addresses this regression.

sjrd committed:
Optimize typical matchEnd() LabelDefs for LoadDestination.

To do this, we intercept typical shapes produced by patmat in
`genBlockTo`.

sjrd committed:
Inliner: Load return values on the stack before jumping to post-call.

Previously, the inliner would emit the following sequence to
replace xRETURN instructions:

  1. STORE the return value
  2. DROP the rest of the stack
  3. JUMP to post-call

And at post-call, it would start by LOADing the return value back
on the stack.

This commit moves (and potentially duplicates) the LOAD instruction
to before JUMPing to post-call. In the common case where the stack
is empty at the point of the xRETURN (save for the return value
itself), we can then completely omit storing and loading the value.

This change recovers the optimized bytecode that the LoadDestination
commit destroyed in `InlinerTest.oldInlineHigherOrderTest(). The
changes in that commit make it much more frequent to have xRETURN
nodes in the middle of a method body, and hence the new inliner
behavior makes more sense.

Note: there already existed a comment in Inliner.inlineCallsite()
that pretended that loading the return value was part of the
replacement sequence for xRETURN. It was basically lying. The new
behavior is aligned with that comment.

lrytz reviewed (approved):
LGTM, thank you for all the follow-ups! And sorry about the lie.

sjrd commented on test/files/neg/inlineMaxSize.scala:
```diff
@@ -5,5 +5,5 @@ class C {
   @inline final def h = g + g + g + g + g + g + g + g + g + g
   @inline final def i = h + h + h + h + h + h + h + h + h + h
-  @inline final def j = i + i + i
+  @inline final def j = i + i + i + i
```
I had to adapt that test to make it even bigger for the warning to still appear. :p

lrytz commented:
Merging so we can see how it goes on the community build.

lrytz commented:
Sure, I'd be happy to have these improvements :)

SethTisue commented:
@sjrd this (and #10025) didn't merge forward onto 2.13 easily; can you submit a forward-port PR?

sjrd commented:
Probably not this week. But maybe next week? IIUC 2.13.9 is going to wait a bit, isn't it?

SethTisue commented:
> IIUC 2.13.9 is going to wait a bit, isn't it?

Correct. 2.12.16 is imminent, but 2.13.9 won't ship until after Scala 3.2.0 final is released. So yes, next week is plenty soon enough.
